Accumulo 2971: improvements to ChangeSecret Tool

milleruntime opened the pull request:
Created a method to verify user running the tool has write permissions to HDFS and modified the tool to update HDFS before making any changes to zookeeper.

ctubbsii commented on server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java:
```diff
@@ -63,12 +66,15 @@ public static void main(String[] args) throws Exception {
-    String instanceId = rewriteZooKeeperInstance(inst, opts.oldPass, opts.newPass);
-    updateHdfs(fs, inst, instanceId);
+    if (!verifyHdfsWritePermission(fs))
+      System.exit(-1);
```
Might be better to print a message first.

Also (and this is perhaps a minor nit), negative exit codes should not be used. They aren't cross-platform. Also, the special exit codes defined in sysexits.h should be avoided, so best to stick with values `[1,63]`. I'd probably just stick with `1` for this (and the other places in this class, which currently use `-1`).

milleruntime replied:
A message telling the user they don't have permissions? That is printed in the catch (AccessControlException ace) in verifyHdfsWritePermission().

Will fix the System.exits.

ctubbsii replied:
Ah, I missed the message in verifyHdfsWritePermission.

milleruntime replied:
Fixed in 69042bd

mjwall commented:
@milleruntime can you make this against the 1.8 branch?  I cherry picked your commits, but had to make a argument final. You should see it when you go to 1.8.

Couple of comments:
- You left a System.exit(1) in the main method.  I'll annotate the diff
- When I removed that System exit, I got the following
`2016-08-04 17:14:43,704 [start.Main] ERROR: Thread 'org.apache.accumulo.server.util.ChangeSecret' died.
java.lang.NoSuchMethodError: org.apache.hadoop.fs.FileSystem.access(Lorg/apache/hadoop/fs/Path;Lorg/apache/hadoop/fs/permission/FsAction;)V
	at org.apache.accumulo.server.util.ChangeSecret.verifyHdfsWritePermission(ChangeSecret.java:172)
	at org.apache.accumulo.server.util.ChangeSecret.main(ChangeSecret.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.accumulo.start.Main$2.run(Main.java:157)
	at java.lang.Thread.run(Thread.java:745)
`  Probably because I was using hadoop 2.4.1 though.  I was hoping to see the message in the IOException in verifyHdfsWritePermission.
- Can we move the verifyAccumuloIsDown and verifyHdfsWritePermission above the opts parsing.  There is no need to prompt for the instance secret if we know it is going to fail.

I am going to move this to 1.8.1, it can be a bug fix.  I was thinking it changed behavior and was trying to push it into 1.8.0.

mjwall commented on server/base/src/main/java/org/apache/accumulo/server/util/ChangeSecret.java:
```diff
@@ -61,14 +64,16 @@ public static void main(String[] args) throws Exception {
-    String instanceId = rewriteZooKeeperInstance(inst, opts.oldPass, opts.newPass);
-    updateHdfs(fs, inst, instanceId);
+    verifyAccumuloIsDown(inst, opts.oldPass);
+    System.exit(1);
```
This should be removed

milleruntime commented:
Yeah I was also thinking the checks should be before the password prompts.
Sounds good.

On Thursday, August 4, 2016, Michael Wall <notifications@github.com> wrote:

> @milleruntime <https://github.com/milleruntime> can you make this against
> the 1.8 branch? I cherry picked your commits, but had to make a argument
> final. You should see it when you go to 1.8.
>
> Couple of comments:
>
>    - You left a System.exit(1) in the main method. I'll annotate the diff
>    - When I removed that System exit, I got the following 2016-08-04
>    17:14:43,704 [start.Main] ERROR: Thread 'org.apache.accumulo.server.util.ChangeSecret'
>    died. java.lang.NoSuchMethodError: org.apache.hadoop.fs.
>    FileSystem.access(Lorg/apache/hadoop/fs/Path;Lorg/apache/
>    hadoop/fs/permission/FsAction;)V at org.apache.accumulo.server.
>    util.ChangeSecret.verifyHdfsWritePermission(ChangeSecret.java:172) at
>    org.apache.accumulo.server.util.ChangeSecret.main(ChangeSecret.java:68)
>    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at
>    sun.reflect.NativeMethodAccessorImpl.invoke(
>    NativeMethodAccessorImpl.java:57) at sun.reflect.
>    DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>    at java.lang.reflect.Method.invoke(Method.java:606) at
>    org.apache.accumulo.start.Main$2.run(Main.java:157) at
>    java.lang.Thread.run(Thread.java:745) Probably because I was using
>    hadoop 2.4.1 though. I was hoping to see the message in the IOException in
>    verifyHdfsWritePermission.
>    - Can we move the verifyAccumuloIsDown and verifyHdfsWritePermission
>    above the opts parsing. There is no need to prompt for the instance secret
>    if we know it is going to fail.
>
> I am going to move this to 1.8.1, it can be a bug fix. I was thinking it
> changed behavior and was trying to push it into 1.8.0.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/accumulo/pull/136#issuecomment-237687465>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALUpGy11cvzGjzr1H4tTSq-UyCYYWQwKks5qclf5gaJpZM4JXeBV>
> .
>

ctubbsii commented:
Even if it checks beforehand, it could still fail when it tries to actually make the change, right?

milleruntime commented:
Yeah if you entered the wrong password an auth error will be thrown.

On Thursday, August 4, 2016, Christopher Tubbs <notifications@github.com>
wrote:

> Even if it checks beforehand, it could still fail when it tries to
> actually make the change, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/accumulo/pull/136#issuecomment-237729204>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALUpG-cwcHsudrtZVNT53eUZwZrzLjxeks5qcoswgaJpZM4JXeBV>
> .
>

ctubbsii commented:
@milleruntime I was thinking more that the filesystem permissions could change between the check and the attempt to make the change.

milleruntime commented:
@mjwall Did you cherry pick these commits? I don't see them in 1.8 branch.
I am currently addressing your comments with a fix to my last commit.

@ctubbsii <https://github.com/ctubbsii> Do you have an idea of how to get
around this filesystem permission change scenario?

On Fri, Aug 5, 2016 at 12:45 PM, Christopher Tubbs <notifications@github.com
> wrote:

> @milleruntime <https://github.com/milleruntime> I was thinking more that
> the filesystem permissions could change between the check and the attempt
> to make the change.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/accumulo/pull/136#issuecomment-237901299>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALUpG22DD8Fx5GhG3Btw5A77Kjpa5GcWks5qc2ixgaJpZM4JXeBV>
> .
>

mjwall commented:
@milleruntime I did not cherry pick since it wasn't ready.

ctubbsii commented:
@milleruntime No. I just wanted to make sure that I properly understood this check. It seems to be a fast-failure check, but not a strict guarantee of future success of the tool. I think that's fine. It's still an improvement.

milleruntime commented:
@mjwall addressed your first comment with 4d575f8

milleruntime commented:
@mjwall your second comment points out that the FileSystem access check I added is dependent on a newer version than yours. The method I used is annotated InterfaceAudience.LimitedPrivate, meaning it shouldn't be used by external applications, so I will replace it with something better.

milleruntime commented:
@mjwall I moved the permissions check before the password prompt. I can't move verifyAccumuloIsDown before the password prompt since it connects to Zookeeper and requires the password.

milleruntime commented:
Ready to go if no more comments. Squashed changes into e23005faeae9
